OpenSimRT/Common/src/SignalProcessing.cpp

```diff
@@ -182,115 +180,6 @@ LowPassSmoothFilter::filter(const LowPassSmoothFilter::Input& input) {
     return output;
 }
 
-/******************************************************************************/
-LowPassSmoothFilterTS::LowPassSmoothFilterTS(const Parameters& parameters)
-        : parameters(parameters), initializationCounter(parameters.memory - 1),
-          newDataReady(false), dataMatrixReady(false) {
-    ENSURE_POSITIVE(parameters.numSignals);
-    // at least 5 slots to define derivatives (5 - 4 > 0)
-    ENSURE_POSITIVE(parameters.memory - 4);
-    ENSURE_POSITIVE(parameters.cutoffFrequency);
-    // if we need time derivatives then we are between [2, M - 2]
-    ENSURE_BOUNDS(parameters.delay, 2, parameters.memory - 2);
-    if (parameters.calculateDerivatives) {
-        ENSURE_BOUNDS(parameters.splineOrder, 1, 7);
-        if (parameters.splineOrder % 2 == 0) {
-            THROW_EXCEPTION(
-                    "spline order should be an odd number between 1 and 7");
-        }
-    }
-
-    // initialize variables
-    N = parameters.numSignals;
-    M = parameters.memory;
-    D = parameters.delay;
-
-    time = Matrix(1, M, 0.0);
-    data = Matrix(N, M, 0.0);
-
-    xRaw = new double[M];
-    xFiltered = new double[M];
-}
-
-LowPassSmoothFilterTS::~LowPassSmoothFilterTS() {
-    // free allocated memory
-    delete[] xRaw;
-    delete[] xFiltered;
-}
-
-void LowPassSmoothFilterTS::updState(
-        const LowPassSmoothFilterTS::Input& input) {
-    {
-        // lock
-        lock_guard<mutex> locker(monitor);
-
-        // shift data column left and set last column as the new data
-        shiftColumnsLeft(Vector(1, input.t), time);
-        shiftColumnsLeft(input.x, data);
-
-        // check if initialized and set ready state
-        if (initializationCounter > 0) {
-            initializationCounter--;
-        } else {
-            dataMatrixReady = true;
-        }
-
-        // set flag for new entry
-        newDataReady = true;
-    }
-
-    // notify after unlocking
-    cond.notify_one();
-}
-
-LowPassSmoothFilterTS::Output LowPassSmoothFilterTS::filter() {
-    std::unique_lock<std::mutex> lock(monitor);
-    cond.wait(lock, [&]() { return dataMatrixReady && newDataReady; });
-    newDataReady = false; // do not filter unless new data are entered in matrix
-
-    // set dt
-    dt = time[0][M - 1] - time[0][M - 2];
```

Comment:
I know this was removed. I think the calculation of dt might have caused the problem. With this implementation, dt can change due to precision errors. Maybe, we can specify dt externally as a parameter. Can you verify this with the old and new MarkerData?

Reply:
Setting only the `dt` in the LowPassSmoothFilter won't do the trick. When pushing to the time matrix using the `shiftColumnsLeft(Vector(1, input.t), time);` you have to push time values that have constant rate. I have tested it and verified that, using the markerData @60Hz, the pipeline works as long as you replace the time values from file with static values that increment in the acquisition function passed in the pipeline parameters. Eg.:
```
    // acquisition function (simulates acquisition from motion)
    auto dataAcquisitionFunction = [&]() -> MotionCaptureInput {
        static int i = 0;
        static double t = 0.0; // <<<<<<<<<<<<<<<<<<< HERE
        
        MotionCaptureInput input;
        // get frame data
        input.IkFrame = InverseKinematics::getFrameFromMarkerData(
                i, markerData, observationOrder, false);
        input.IkFrame.t = t;

        // get grf force
        auto grfRightWrench = ExternalWrench::getWrenchFromStorage(
                t, grfRightLabels, grfMotion);
        auto grfLeftWrench = ExternalWrench::getWrenchFromStorage(
                t, grfLeftLabels, grfMotion);
        input.ExternalWrenches = {grfRightWrench, grfLeftWrench};

        // dummy delay to simulate real time
        this_thread::sleep_for(10ms);
        i++;
        t += 1 / 60.0; // <<<<<<<<<<<<<<<<<<< HERE
        return input;
    };
```

I think resampling the original data is more appropriate than error-proofing the filter in this case, since data are retrieved from file. In real-time scenarios, one can use the SyncManager (or any other future implementation) to resample the streamed data online.